Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/intersection.h

```diff
@@ -26,9 +26,769 @@
+
+  VertexPointMap vpmap = boost::choose_param(get_param(np, internal_np::vertex_point),
+                                             get_const_property_map(boost::vertex_point, mesh));
+  typedef typename boost::property_traits<VertexPointMap>::value_type Point;
+
+  std::vector<face_descriptor> faces;
+  faces.reserve(std::distance( boost::begin(face_range), boost::end(face_range) ));
+
+  typedef typename CGAL::Box_intersection_d::Box_with_info_d<double, 3, std::size_t> Box;
+
+  // make one box per facet
+  std::vector<Box> boxes1;
+  std::vector<Box> boxes2;
+  boxes1.reserve(
+        std::distance( boost::begin(face_range), boost::end(face_range) )
+        );
+
+  boxes2.reserve(
+        std::distance( boost::begin(polyline), boost::end(polyline) ) - 1
+        );
+
+
+  BOOST_FOREACH(face_descriptor f, face_range)
+  {
+    faces.push_back(f);
+    boxes1.push_back(Box(Polygon_mesh_processing::face_bbox(f, mesh), faces.size()-1));
+  }
+
+  for(std::size_t i =0; i< polyline.size()-1; ++i)
+  {
+    Point p1 = polyline[i];
+    Point p2 = polyline[i+1];
+    boxes2.push_back(Box(p1.bbox() + p2.bbox(), i));
+  }
+
+  // generate box pointers
+  std::vector<const Box*> box1_ptr;
+  std::vector<const Box*> box2_ptr;
+  box1_ptr.reserve(boxes1.size());
+  box2_ptr.reserve(boxes2.size());
+
+  BOOST_FOREACH(Box& b, boxes1)
+      box1_ptr.push_back(&b);
+  BOOST_FOREACH(Box& b, boxes2)
+      box2_ptr.push_back(&b);
+
+  // compute intersections filtered out by boxes
+  typedef typename GetGeomTraits<TM, NamedParameters>::type GeomTraits;
+
+  CGAL::internal::Intersect_face_polyline<TM,
+      GeomTraits,
+      Box,
+      OutputIterator,
+      VertexPointMap>
+      Intersect_face_polyline(mesh,
+                               faces,
+                               polyline,
+                               out,
+                               vpmap,
+                               GeomTraits());
+
+  std::ptrdiff_t cutoff = 2000;
+  CGAL::box_intersection_d(box1_ptr.begin(), box1_ptr.end(),
+                           box2_ptr.begin(), box2_ptr.end(),
+                           Intersect_face_polyline, cutoff);
+  return Intersect_face_polyline.m_iterator;
+}
+
+/*
+ * \ingroup PMP_intersection_grp
+ * detects and records intersections between two `Polyline`s.
+ * This function depends on the package \ref PkgBoxIntersectionDSummary
+ *
+ * \tparam Polyline a `RandomAccessRange` of `CGAL::Point_3<Kernel>`.
+ * \tparam OutputIterator a model of `OutputIterator` holding objects of type
+ *   `std::pair<std::size_t, std::size_t>`. This OutputIterator will hold the position of the
+ *  elements in their respective range.
+ * \tparam Kernel a model of `Kernel`
+ *
+ * \param polyline1 the first polyline to check for intersections.
+ * \param polyline2 the second polyline to check for intersections.
+ * \param out output iterator to be filled with all pairs of segments that intersect
```

Comment:
`to check` vs `to be filled` I don't know which one is correct but I guess the same formulation should be used in both cases.